projects/baby_driver/test/test_gpio_interrupts.c

```diff
@@ -0,0 +1,280 @@
+                                                          bool *tx_result) {
+  s_times_register_callback_called++;
+  s_received_status = data[1];
+  s_received_context = context;
+  *tx_result = false;
+  return STATUS_CODE_OK;
+}
+
+// Called when the interrupt handler is called
+static StatusCode prv_rx_gpio_interrupt_handler_callback(uint8_t data[8], void *context,
+                                                         bool *tx_result) {
+  s_times_handler_callback_called++;
+  s_test_port = data[1];
+  s_test_pin = data[2];
+  s_test_edge = data[3];
+  s_received_context = context;
+  *tx_result = false;
+  return STATUS_CODE_OK;
+}
+
+void setup_test(void) {
+  initialize_can_and_dependencies(&s_can_storage, SYSTEM_CAN_DEVICE_BABYDRIVER, TEST_CAN_EVENT_TX,
+                                  TEST_CAN_EVENT_RX, TEST_CAN_EVENT_FAULT);
+  interrupt_init();
+  soft_timer_init();
+  gpio_init();
+  gpio_it_init();
+  TEST_ASSERT_OK(dispatcher_init());
+  TEST_ASSERT_OK(gpio_interrupts_init());
+  s_times_register_callback_called = 0;
+  s_times_handler_callback_called = 0;
+  s_received_status = 0;
+  s_test_port = 0;
+  s_test_pin = 0;
+  s_test_edge = 0;
+
+  TEST_ASSERT_OK(dispatcher_register_callback(BABYDRIVER_MESSAGE_STATUS,
+                                              prv_rx_register_gpio_interrupt_callback, NULL));
+  TEST_ASSERT_OK(dispatcher_register_callback(BABYDRIVER_MESSAGE_GPIO_IT_INTERRUPT,
+                                              prv_rx_gpio_interrupt_handler_callback, NULL));
+}
+
+void teardown_test(void) {}
+
+// Test that gpio interrupts are registered, unregistered, and reregistered correctly
+void test_register_gpio_interrupts(void) {
+  GpioPort port = GPIO_PORT_A;
+  uint8_t pin = 15;
+  InterruptEdge edge = INTERRUPT_EDGE_RISING;
+  // Since edge is rising gpio state will become high
+  s_returned_state = GPIO_STATE_HIGH;
+  GpioAddress address = { .port = port, .pin = pin };
+  uint8_t data[8] = { BABYDRIVER_MESSAGE_GPIO_IT_REGISTER_COMMAND, port, pin, edge, 0, 0, 0, 0 };
+
+  // Test whether a single register command works
+  // Send CAN message with register command message information
+  CAN_TRANSMIT_BABYDRIVER(data[0], data[1], data[2], data[3], data[4], data[5], data[6], data[7]);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(1, s_times_register_callback_called);
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, s_received_status);
+
+  // Trigger the gpio interrupt and a message should be received
+  gpio_it_trigger_interrupt(&address);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(1, s_times_handler_callback_called);
+  // Compares data from command message with data transmitted after the gpio interrupt was triggered
+  TEST_ASSERT_EQUAL(port, s_test_port);
+  TEST_ASSERT_EQUAL(pin, s_test_pin);
+  TEST_ASSERT_EQUAL(edge, s_test_edge);
+
```

Comment:
Minor additions to this test: you should test that the CAN message is not received when an interrupt is triggered on the wrong edge (e.g. a falling interrupt here).